src/server/scripts/Northrend/Naxxramas/boss_kelthuzad.cpp

```diff
@@ -268,355 +144,218 @@ class boss_kelthuzad : public CreatureScript
-                            events.ScheduleEvent(EVENT_BLAST, urand(60000, 120000));
-                            if (GetDifficulty() == RAID_DIFFICULTY_25MAN_NORMAL)
-                                events.ScheduleEvent(EVENT_CHAIN, urand(30000, 60000));
-                            events.SetPhase(PHASE_TWO);
-                            break;
-                        default:
-                            break;
+                        events.ScheduleEvent(EVENT_TRANSITION_SUMMON, urandms(19,21), 0, PHASE_TWO);
+                        events.ScheduleEvent(EVENT_TRANSITION_SUMMON, urandms(25,27), 0, PHASE_TWO);
                     }
                 }
-            }
-            else
-            {
-                if (me->HasUnitState(UNIT_STATE_CASTING))
-                    return;
 
-                if (uint32 eventId = events.ExecuteEvent())
+                while (uint32 eventId = events.ExecuteEvent())
                 {
+                    if (_frostboltCooldown <= 4 * IN_MILLISECONDS) // stop casting bolts for 4 seconds after doing another action
+                        _frostboltCooldown = 4 * IN_MILLISECONDS;
                     switch (eventId)
                     {
-                        case EVENT_BOLT:
-                            DoCastVictim(SPELL_FROST_BOLT);
-                            events.Repeat(5000, 10000);
+                        case EVENT_PHASE_TWO:
+                            me->CastStop();
+                            events.SetPhase(PHASE_TWO);
+                            me->RemoveFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_NOT_SELECTABLE | UNIT_FLAG_IMMUNE_TO_PC);
+                            me->getThreatManager().resetAllAggro();
+                            me->SetReactState(REACT_AGGRESSIVE);
+                            Talk(EMOTE_PHASE_TWO);
+
+                            _frostboltCooldown = 2 * IN_MILLISECONDS;
+                            events.ScheduleEvent(EVENT_FROSTBOLT_VOLLEY, urandms(36,44), 0, PHASE_TWO);
+                            events.ScheduleEvent(EVENT_SHADOW_FISSURE, urandms(6, 10), 0, PHASE_TWO);
+                            events.ScheduleEvent(EVENT_DETONATE_MANA, urandms(27, 33), 0, PHASE_TWO);
+                            events.ScheduleEvent(EVENT_FROST_BLAST, urandms(20, 25), 0, PHASE_TWO); // @todo sniff timer
+                            if (Is25ManRaid())
+                                events.ScheduleEvent(EVENT_CHAINS, urandms(60, 80), 0, PHASE_TWO); // @todo sniff timer
                             break;
-                        case EVENT_NOVA:
-                            DoCastAOE(SPELL_FROST_BOLT_AOE);
-                            events.Repeat(15000, 30000);
+                        
+                        case EVENT_FROSTBOLT_VOLLEY:
+                            DoCastAOE(SPELL_FROSTBOLT_VOLLEY);
+                            if (!urand(0, 3)) // sometimes he skips one volley - is this truly random? i couldn't discern any pattern
```

Comment:
I assume this may be caused by the spell being locked by category(some used here uses category 1152), i never understood how this is handled by blizz